modules/core/src/main/scala/sangria/schema/SchemaValidationRule.scala

```diff
@@ -410,6 +411,31 @@ object EnumValueReservedNameValidator extends SchemaElementValidator {
     else Vector.empty
 }
 
+object InputObjectTypeRecursionValidator extends SchemaElementValidator {
+  override def validateInputObjectType(schema: Schema[_, _], tpe: InputObjectType[_]): Vector[Violation] = {
+    containsRecursiveInputObject(tpe.namedType.name, List(), schema, tpe)
+  }
+
+  private def containsRecursiveInputObject(rootTypeName: String, path: List[String], schema: Schema[_, _], tpe: InputObjectType[_]): Vector[Violation] = {
+    val recursiveFields = tpe.fields.filter(childField => childField.fieldType.namedType.name == rootTypeName && !childField.fieldType.isOptional && !childField.fieldType.isList)
+    if (recursiveFields.nonEmpty) {
+      recursiveFields.flatMap(field => Vector(InputObjectTypeRecursion(tpe.name, field.name, path, None, Nil))).toVector
+    } else {
+      var violations = Vector[Violation]()
+      val childTypesToCheck = tpe.fields.filter(field => !field.fieldType.isOptional && !field.fieldType.isList && field.fieldType.isInstanceOf[InputObjectType[_]])
+      childTypesToCheck.foreach { field =>
+        schema.getInputType(NotNullType(NamedType(field.fieldType.namedType.name))).asInstanceOf[Option[InputObjectType[_]]] match {
+          case Some(objectType) if objectType != tpe =>
+            val updatedPath = path :+ field.name
+            violations = violations ++ containsRecursiveInputObject(rootTypeName, updatedPath, schema, objectType)
+          case _ =>
+        }
+      }
+      violations
```

Comment:
To avoid the `var`:
```suggestion
      val childTypesToCheck = tpe.fields.filter(field => !field.fieldType.isOptional && !field.fieldType.isList && field.fieldType.isInstanceOf[InputObjectType[_]])
      childTypesToCheck.foldLeft(Vector.empty[Violation]) { case (acc, field) =>
        schema.getInputType(NotNullType(NamedType(field.fieldType.namedType.name))) match {
          case Some(objectType: InputObjectType[_]) if objectType != tpe =>
            val updatedPath = path :+ field.name
            acc ++ containsRecursiveInputObject(rootTypeName, updatedPath, schema, objectType)
          case _ => acc
        }
      }
```

Reply:
👍